rust/diesel/src/order.rs

```diff
@@ -0,0 +1,176 @@
+use bigdecimal::BigDecimal;
+use diesel;
+use diesel::pg::PgConnection;
+use diesel::prelude::*;
+use num_traits::identities::Zero;
+use serde::{Deserialize, Serialize};
+use std::ops::{Add, Mul};
+
+use crate::product::Product;
+use crate::schema::order_lines;
+use crate::schema::orders;
+
+#[derive(AsChangeset, Serialize, Deserialize, Queryable, Insertable)]
+#[serde(rename_all = "camelCase")]
+pub struct Order {
+    pub order_id: i32,
+    pub order_total: BigDecimal,
+    pub user_id: i32,
+}
+
+#[table_name = "orders"]
+#[derive(AsChangeset, Serialize, Deserialize, Queryable, Insertable)]
+#[serde(rename_all = "camelCase")]
+struct NewOrder {
+    pub order_total: BigDecimal,
+    pub user_id: i32,
+}
+
+#[table_name = "order_lines"]
+#[derive(AsChangeset, Serialize, Deserialize, Queryable, Insertable)]
+#[serde(rename_all = "camelCase")]
+pub struct NewOrderLine {
+    pub order_id: i32,
+    pub product_id: i32,
+    pub units: i16,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct NewUserOrder {
+    pub user_id: i32,
+    pub products: Vec<ProductOrder>,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ProductOrder {
+    pub product_id: i32,
+    pub units: i16,
+}
+
+impl Order {
+    pub fn create(order: NewUserOrder, connection: &PgConnection) -> Option<Order> {
+        let product_units: Vec<(Product, i16)> = order
+            .products
+            .iter()
+            .flat_map(|product_ref| {
+                Product::read(product_ref.product_id, connection)
+                    .map(|product| (product, product_ref.units))
+            })
+            .collect();
+
+        let order_total = product_units
+            .clone()
+            .iter()
+            .fold(BigDecimal::from(0.0), |acc, (product, qty)| {
+                acc.add(BigDecimal::from(*qty).mul(&product.price))
+            });
+
+        if order_total.is_zero() {
+            return None;
+        }
+
+        let row = NewOrder {
+            user_id: order.user_id,
+            order_total,
+        };
+
+        let insert_result = connection
+            .build_transaction()
+            .serializable()
+            .read_write()
+            .run(|| {
+                let inserted_order: Order = diesel::insert_into(orders::table)
+                    .values(&row)
+                    .get_result(connection)?;
+
+                let order_lines: Vec<NewOrderLine> = product_units
+                    .clone()
```

Comment:
Shouldn't need to clone here (or in the map below), just do the same thing as in the update method.